src/arch/x86/x86_64/src/bios.rs

```diff
@@ -0,0 +1,380 @@
+    pub xgpe1_block: AcpiGenericAddress,         /* 64-bit Extended General Purpose Event 1 Reg Blk address */
+    pub sleep_control: AcpiGenericAddress,       /* 64-bit Sleep Control register (ACPI 5.0) */
+    pub sleep_status: AcpiGenericAddress,        /* 64-bit Sleep Status register (ACPI 5.0) */
+    pub hypervisor_id: u64,                      /* Hypervisor Vendor ID (ACPI 6.0) */
+}
+
+#[repr(packed)]
+#[derive(Default)]
+pub struct AcpiTableMadt {
+    pub header: AcpiTableHeader, /* Common ACPI table header */
+    pub address: u32,            /* Physical address of local APIC */
+    pub flags: u32,
+}
+
+#[repr(packed)]
+#[derive(Default)]
+pub struct AcpiSubtableHader {
+    pub r#type: u8,
+    pub length: u8,
+}
+
+#[repr(packed)]
+#[derive(Default)]
+pub struct AcpiMadtLocalApic {
+    pub header: AcpiSubtableHader,
+    pub processor_id: u8, /* ACPI processor id */
+    pub id: u8,           /* Processor's local APIC id */
+    pub lapic_flags: u32,
+}
+
+#[repr(packed)]
+#[derive(Default)]
+pub struct AcpiMadtIoApic {
+    pub header: AcpiSubtableHader,
+    pub id: u8,               /* I/O APIC ID */
+    pub reserved: u8,         /* reserved - must be zero */
+    pub address: u32,         /* APIC physical address */
+    pub global_irq_base: u32, /* Global system interrupt where INTI lines start */
+}
+
+#[repr(packed)]
+#[derive(Default)]
+pub struct AcpiMadtLocalX2apic {
+    pub header: AcpiSubtableHader,
+    pub reserved: u16,      /* reserved - must be zero */
+    pub local_apic_id: u32, /* Processor x2APIC ID  */
+    pub lapic_flags: u32,
+    pub uid: u32, /* ACPI processor UID */
+}
+
+#[repr(packed)]
+#[derive(Default)]
+pub struct AcpiMadtInterruptOverride {
+    pub header: AcpiSubtableHader,
+    pub bus: u8,
+    pub sourceirq: u8,
+    pub globalirq: u32,
+    pub flags: u16,
+}
+
+#[repr(packed)]
+#[derive(Default)]
+pub struct AcpiMadtNMI {
+    pub header: AcpiSubtableHader,
+    pub flags: u16,
+    pub globalirq: u32,
+}
+
+#[repr(packed)]
+#[derive(Default)]
+pub struct AcpiMadtLAPICNMI {
+    pub header: AcpiSubtableHader,
+    pub acpi_processor_uid: u8,
+    pub flags: u16,
+    pub local_interrupt: u8,
+}
+
+fn write<T>(w: &mut print::WriteTo, val: T, offset: usize, index: usize) {
+    let y = (offset + index * size_of::<T>()) as *mut T;
+    unsafe {
+        ptr::write_volatile(y, val);
+    }
+    write!(w, "write to {:x?}: \r\n", y).unwrap();
+}
+
+fn read<T>(offset: usize, index: usize) -> T {
+    let y = (offset + index * size_of::<T>()) as *mut T;
+    let v = unsafe { ptr::read_volatile(y) };
+    v
+}
+
+/*
+ *
+ * Intel ACPI Component Architecture
+ * ASL Optimizing Compiler version 20140214-64 [Mar 29 2014]
+ * Copyright (c) 2000 - 2014 Intel Corporation
+ *
+ * Compilation of "vmm_acpi_dsdt.dsl" - Fri Apr  1 13:34:26 2016
+ *
+ */
+
+/*
+ *
+ * Based on the example at osdev wiki wiki.osdev.org/AML,
+ * and the other example in http://www.acpi.info/DOWNLOADS/ACPI_5_Errata%20A.pdf
```

Comment:
Could you please add https://uefi.org/acpi/specs to save people some time digging? Things got moved around and the former ACPI website now just redirects to Intel. 🤡